engine/table/src/main/java/io/deephaven/engine/table/impl/TimeTable.java

```diff
@@ -33,37 +47,47 @@
 
     private static final String TIMESTAMP = "Timestamp";
     private long lastIndex = -1;
-    private final DateTimeArraySource dateTimeArraySource;
+    private final SyntheticDateTimeSource columnSource;
     private final TimeProvider timeProvider;
-    private DateTime lastTime;
-    private final long period;
-    private final long binOffset;
     private final PerformanceEntry entry;
+    private final boolean isStreaming;
 
     public TimeTable(TimeProvider timeProvider, long period) {
-        this(timeProvider, null, period);
+        this(timeProvider, null, period, false);
+    }
+
+    public TimeTable(TimeProvider timeProvider, long period, boolean isStreaming) {
+        this(timeProvider, null, period, isStreaming);
     }
 
     public TimeTable(TimeProvider timeProvider, DateTime firstTime, long period) {
-        super(RowSetFactory.fromKeys().toTracking(), initColumn());
+        this(timeProvider, firstTime, period, false);
+    }
+
+    public TimeTable(TimeProvider timeProvider, DateTime firstTime, long period, boolean isStreaming) {
+        super(RowSetFactory.fromKeys().toTracking(), initColumn(period));
         if (period <= 0) {
             throw new IllegalArgumentException("Invalid time period: " + period + " nanoseconds");
         }
-        this.entry = UpdatePerformanceTracker.getInstance().getEntry("TimeTable(" + firstTime + "," + period + ")");
-        this.lastTime = firstTime == null ? null : new DateTime(firstTime.getNanos() - period);
-        binOffset = firstTime == null ? 0 : lastTime.getNanos() % period;
-        dateTimeArraySource = (DateTimeArraySource) getColumnSourceMap().get(TIMESTAMP);
+        this.isStreaming = isStreaming;
+        final String name = isStreaming ? "TimeTableStream" : "TimeTable";
+        this.entry = UpdatePerformanceTracker.getInstance().getEntry(name + "(" + firstTime + "," + period + ")");
+        columnSource = (SyntheticDateTimeSource) getColumnSourceMap().get(TIMESTAMP);
+        columnSource.baseTime = firstTime;
```

Comment:
Couldn't we pass this into the column source constructor along with `period`? Better, we could pass `timeProvider` in, too, and always initialize `baseTime` in the constructor, allowing it to be `final`.

Reply:
As discussed on slack, initializing the base time would change the semantics slightly; this is currently deferred to the first refresh.